File: Image/Extensions/Layer/Layer.cpp
#include <vector>
using namespace std;
namespace Tahsin
{
    template <typename T>
    class Layer
    {
    protected:
        vector<vector<T>> data;
        vector<vector<int>> permission;

    public:
        vector<T> dataVarierty;
        Layer(int width, int height, int permissionCount)
        {
            for (int i = 0; i < height; i++)
            {
                vector<T> temp(width);
                data.push_back(temp);
            }
        }
        virtual void GenerateData() = 0;
        virtual vector<vector<T>> GetData() = 0;
    };
}

File: Image/Extensions/Layer/Layer.cs
namespace TahsinsLibrary.Layers
{
    public abstract class Layer<T>
    {
        public T[] dataVarierty;
        protected int[,] permission;
        protected T[,] data;
        public Layer(int width, int height, int permissionCount)
        {
            data = new T[width, height];
            permission = new int[width, height];
            for (int i = 0; i < width; i++)
            {
                for (int j = 0; j < height; j++)
                {
                    permission[i, j] = permissionCount;
                }
            }
        }
        public abstract void GenerateData();
        public abstract T[,] GetData();
    }
}